testutils/util/artifact.py

```diff
@@ -0,0 +1,206 @@
+
+    def _complete_manifest(self):
+        self._artifact.seek(0)
+        rdtar = tarfile.open(fileobj=self._artifact, mode="r")
+        manifest_hdr = rdtar.getmember("manifest")
+        offset = manifest_hdr.offset_data
+        for filename in self._filenames[::-1]:
+            offset_sha = offset + self._manifest_offsets[filename]
+            self._artifact.seek(offset_sha)
+            self._artifact.write(self._shasums[filename].encode())
+
+    def _add_payloads(self):
+        """
+        Adds all the stored payload to artifact.
+        Each payload is itself a compressed tar.
+        """
+        filenames = sorted(list(self._payloads.keys()))
+        for filename in filenames:
+            fd = self._payloads[filename]
+
+            size = fd.seek(0, io.SEEK_END)
+            fd.seek(0)
+
+            payload_tarbin = io.BytesIO()
+            payload_tar = tarfile.open(fileobj=payload_tarbin, mode="w:gz")
+            tarhdr = tarfile.TarInfo(os.path.basename(filename))
+            tarhdr.size = size
+            payload_tar.addfile(tarhdr, fd)
+            payload_tar.close()
+
+            tarhdr = tarfile.TarInfo(os.path.dirname(filename) + ".tar.gz")
+            tarhdr.size = payload_tarbin.tell()
+            self._compute_checksum(filename, fd)
+            payload_tarbin.seek(0)
+            self._tarfact.addfile(tarhdr, payload_tarbin)
+
+    def _add_version(self):
+        version = {"format": "mender", "version": 3}
+        fd = io.BytesIO(json.dumps(version).encode())
+        size = self._compute_checksum("version", fd)
+        tarhdr = tarfile.TarInfo("version")
+        tarhdr.size = size
+        self._tarfact.addfile(tarhdr, fd)
+
+    def _add_header(self):
+        hdr_tarbin = io.BytesIO()
+        hdr_tar = tarfile.open(fileobj=hdr_tarbin, mode="w:gz")
+        header_info = {
+            "payloads": [
+                {
+                    "type": self._payload_types[filename]
+                    for filename in sorted(self._payloads.keys())
+                }
+            ]
+        }
+        header_info["artifact_provides"] = self._provides["header-info"]
+        header_info["artifact_depends"] = self._depends["header-info"]
+        hdr_info = io.BytesIO(json.dumps(header_info).encode())
+        size = hdr_info.seek(0, io.SEEK_END)
+        hdr_info.seek(0)
+        tarhdr = tarfile.TarInfo("header-info")
+        tarhdr.size = size
+        hdr_tar.addfile(tarhdr, hdr_info)
+
+        for filename in sorted(self._payloads.keys()):
+            path_prefix = os.path.join(
+                "headers", os.path.basename(os.path.dirname(filename))
+            )
+            typeinfo = {"type": self._payload_types[filename]}
+            if filename in self._depends:
+                typeinfo["artifact_depends"] = self._depends[filename]
+            if filename in self._provides:
+                typeinfo["artifact_provides"] = self._provides[filename]
```

Comment:
This could override an existing depends/provides, but I guess that is the correct behavior. Does mender-artifact give an error in these cases? Could we add some error here?

Reply:
That's true. I think we should update the artifact v3 specification as well as it doesn't say anything that the `type-info` provides/depends has to be globally unique.